# Error management

## 7.1 Panicking

## 7.2 Ignoring when to wrap an error

Error wrapping is about wrapping or packing up an error inside a wrapper container that also make the source error avaiable.

```go
type badError struct {
    err error
}

func (b *badError) Error() string {
    return "bad error, " + err.Error()
}

func main() {
    err := badError{err: errors.New("bad error")}
    wrapperErr := fmt.Errorf("wrapper: %w", err)
    fmt.Println(errors.Unwrap(wrapperError).(*badError).err)
}
```
## 7.3 Comparing an error type inaccurately

```go
type badError struct {
	err error
}

func (b *badError) Error() string {
	return fmt.Sprintf("bad error: %s", b.err.Error())
}

func main() {
	var err error = &badError{
		err: errors.New("db connection"),
	}
	err1 := fmt.Errorf("failed: %w", err)
	err2 := fmt.Errorf("failed: %w", err1)
	var bErr *badError
	ok := errors.As(err2, &bErr)
	fmt.Println(ok)
}
```

## 7.4 Comparing an error value inaccurately

If we use in our application error wrapping using the %w directive and fmt.Errorf, checking an error against a specific value shouldn't be done using == but using errors.Is. Thus, even if the sentinel error is wrapped, errors.Is will be able to recursively unwrap it and compare each error in the chain against the provided value.

## 7.5 Handling an error twice

Handling an error should be done only once. Logging an error is handling an error. Hence, we should either log or return an error. By doing this, we simplify our code and gain better insights into an error situation. Using error wrapping is the most convenient approach as it allows us to propagate the source error and add context to an error.

## 7.6 Not handling an error

Ignoring an error in Go should be execeptional. In many cases, we may still favor logging them, even at a low level log level. Yet, if we are sure that an error can and should be ignored, we must do it explicitly by assigning it to the blank identifier. This way, a future reader will understand it was intentional.

## 7.7 Not handling defer errors

```go
func main() {
    defer func(){
        _ = rows.Close()
    }()
}
```

## 7.8 Summary

- Using panic is an option to deal with errors in Go.
- Error wrapping with the %w directive and fmt.Errorf, comparing an error against a type or a value has to be done respectively using errors.As or errors.Is
